Setup client deployment

florimondmanca opened the pull request:
Closes #46 

Cette PR étend la configuration Ansible pour déployer le client sur le serveur (en même temps que l'API), incluant :

* Installation de Node lors du provisioning, installation des dépendances Node et lancée du build lors du deploy
* Ajout du serveur Node frontend à Supervisor et Nginx
* Paramétrisation de l'URL lors de requêtes à l'API (avec un setup assez subtil pour faire correspondre SSR/browser x prod/dev)
* Mise à jour de la doc

Testé en local avec la VM de test, déploiement sur staging à venir.

magopian reviewed (approved):
Je n'y connais pas grand chose à Ansible, mais après avoir fait une passe de relecture, ça m'a l'air solide 👍 
Merci !

🚢

florimondmanca commented on client/svelte.config.js:
```diff
@@ -17,6 +19,22 @@ const config = {
+              proxy: {
+                // Make the location of the API server in production (<domain>/api) available during development.
+                "/api": {
+                  target: "http://localhost:3579",
```
Oui, pourquoi pas.

Ça demande un peu de config en plus puisqu'alors on voudrait j'imagine rendre le port de l'API configurable aussi. Mais à la compilation le client n'a accès qu'aux variables d'env de type `import.meta.env.VITE_...` (évite de leak des variables d'env dans le code client).

Donc je vais sûrement faire une PR à part pour ça, et en profiter pour renommer les variables `APP_...` en `API_...` (et donc par ex `API_PORT`).

florimondmanca commented on client/jest.config.cjs:
```diff
@@ -10,10 +10,7 @@ module.exports = {
-      "<rootDir>/.svelte-kit/dev/runtime/app$1",
-      "<rootDir>/.svelte-kit/build/runtime/app$1",
-    ],
+    "^\\$app(.*)$": "<rootDir>/src/__tests__/appMock.cjs",
```
Je ne suis pas sûr de comprendre moi-même ce qui faisait que l'état initial ne fonctionnait pas...

Sans ce changement, quand on lance `make test` on obtient l'erreur suivante, qui semble similaire à https://github.com/rossyman/svelte-add-jest/issues/12 :

```console
$ make test-client
cd client && npm run test && npm run test:coverage

> client@0.0.1 test
> jest

 FAIL  src/__tests__/index.test.ts
  ● Test suite failed to run

    Configuration error:
    
    Could not locate module $app/env mapped as:
    [
      "/home/florimond/dev/prj-catalogue/catalogage-donnees/client/.svelte-kit/dev/runtime/app/env",
      "/home/florimond/dev/prj-catalogue/catalogage-donnees/client/.svelte-kit/build/runtime/app/env"
    ].
    
    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/^\$app(.*)$/": "[
          "/home/florimond/dev/prj-catalogue/catalogage-donnees/client/.svelte-kit/dev/runtime/app$1",
          "/home/florimond/dev/prj-catalogue/catalogage-donnees/client/.svelte-kit/build/runtime/app$1"
        ]"
      },
      "resolver": undefined
    }

    > 1 | import { browser } from "$app/env";
        | ^
      2 |
      3 | export const getApiUrl = () => {
      4 |   if (!browser) {

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/resolver.js:579:17)
      at Object.<anonymous> (src/lib/fetch.ts:1:1)
```

En effet les dossiers `.svelte-kit/(dev|build)` n'existent pas / ne sont pas créés... De mon côté j'ai seulement `.svelte-kit/runtime` et `.svelte-kit/generated`.

Je n'ai pas trouvé dans la doc (parcellaire) de SvelteKit à quoi ils servaient ni dans quel contexte ils pourraient être créés.

Peut-être que le setup SvelteKit/Jest par défaut n'a pas encore pris en compte un changement récent dans SvelteKit (qui est en version `next`, tout de même).

Alors j'ai essayé de pointer vers `.svelte-kit/runtime`, vu que lui au moins on l'a.

Mais alors j'obtenais un autre soucis, qui semblait cette fois lié à l'intégration avec Vite (le bundler utilisé par SvelteKit), notamment cette issue https://github.com/rossyman/svelte-add-jest/issues/14 :

```console
$ make test-client
cd client && npm run test && npm run test:coverage

> client@0.0.1 test
> jest

 FAIL  src/__tests__/index.test.ts
  ● Test suite failed to run

    .svelte-kit/runtime/app/env.js:6:18 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node12', or 'nodenext'.

    6 const browser = !import.meta.env.SSR;
                       ~~~~~~~~~~~
    .svelte-kit/runtime/app/env.js:10:15 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node12', or 'nodenext'.

    10 const dev = !!import.meta.env.DEV;
                     ~~~~~~~~~~~
    .svelte-kit/runtime/app/env.js:14:14 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node12', or 'nodenext'.

    14 const mode = import.meta.env.MODE;
                    ~~~~~~~~~~~
    .svelte-kit/runtime/app/env.js:18:15 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node12', or 'nodenext'.

    18 const amp = !!import.meta.env.VITE_SVELTEKIT_AMP;
                     ~~~~~~~~~~~

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        2.665 s, estimated 3 s
Ran all test suites.
make: *** [Makefile:57: test-client] Error 1
```

Comme cette entrée de `moduleNameMapper` sert à dire à Jest (la lib de test) "voici ce que tu dois lire quand tu vois un import de `$app/*`", et qu'actuellement on ne fait que récupérer `import {browser} from "$app/env"`, j'ai fait ce truc un peu sale : pointer vers un module qui définit `browser` (en réalité on n'a pas besoin du `dev`, je vais le retirer). Sachant qu'à ce stade en réalité la présence / valeur de `browser` n'importe pas au vu des tests que l'on fait.

Bref, voilà, le bazar des outils de compilation JS que j'ai juste essayé de faire fonctionner, en gros ^^

florimondmanca replied:
J'en parlerai à l'occasion avec le M. SvelteKit à Fairness, probablement vendredi.

magopian replied:
#fatigue

Merci pour l'explication, et du coup ça me convient en l'état (et je suis curieux du retour de M. SvelteKit ;)

magopian commented:
Pour info il y a un soucis de formatting sur `src/__tests__/appMock.cjs` (si tu lances `npm run format` ça devrait te corriger ça tout seul comme un grand).

florimondmanca commented:
# Test sur staging

Provisioning (installation de Node, etc)

```console
$ make provision-staging
cd ops && make provision-staging
make[1]: Entering directory '/home/florimond/dev/prj-catalogue/catalogage-donnees/ops'
../venv/bin/ansible-playbook -i ansible/hosts_staging -e @ansible/secrets/staging.enc --vault-password-file ansible/secrets/staging.vault-password -t provision ansible/playbook.yml

PLAY [web] *********************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************
ok: [web1-staging]

TASK [common : Update system] **************************************************************************************************************************************************************************
changed: [web1-staging]

TASK [pyenv : Install build requirements] **************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Clone pyenv] *****************************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Set path to pyenv binary] ****************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Create .pyenvrc] *************************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Set path to .bashrc] *********************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Load pyenv in .bashrc] *******************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Install interpreter] *********************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Get current global version] **************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Set pyenv_global_active fact] ************************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Check if 'system' version is available] **************************************************************************************************************************************************
ok: [web1-staging]

TASK [pyenv : Remove 'system' from pyenv_global] *******************************************************************************************************************************************************
skipping: [web1-staging]

TASK [pyenv : Set pyenv global version to '3.8.9 system'] **********************************************************************************************************************************************
skipping: [web1-staging]

TASK [nvm : Install nvm] *******************************************************************************************************************************************************************************
changed: [web1-staging]

TASK [nvm : Install Node binary] ***********************************************************************************************************************************************************************
changed: [web1-staging]

TASK [nvm : Get node binary path] **********************************************************************************************************************************************************************
changed: [web1-staging]

TASK [nvm : Set node_bin_path fact] ********************************************************************************************************************************************************************
ok: [web1-staging]

TASK [nvm : Get npm binary path] ***********************************************************************************************************************************************************************
changed: [web1-staging]

TASK [nvm : Set npm_bin_path fact] *********************************************************************************************************************************************************************
ok: [web1-staging]

TASK [nvm : Create node bin symlink] *******************************************************************************************************************************************************************
changed: [web1-staging]

TASK [nvm : Create npm bin symlink] ********************************************************************************************************************************************************************
changed: [web1-staging]

TASK [common : Ensure nginx is installed and up to date] ***********************************************************************************************************************************************
ok: [web1-staging]

TASK [common : Ensure supervisor is installed] *********************************************************************************************************************************************************
ok: [web1-staging]

PLAY RECAP *********************************************************************************************************************************************************************************************
web1-staging               : ok=22   changed=7    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   

make[1]: Leaving directory '/home/florimond/dev/prj-catalogue/catalogage-donnees/ops'
```

Deploy :

```console
$ EXTRA_OPTS="-e env_branch=fm/client-ops" make deploy-staging
cd ops && make deploy-staging
make[1]: Entering directory '/home/florimond/dev/prj-catalogue/catalogage-donnees/ops'
../venv/bin/ansible-playbook -i ansible/hosts_staging -e @ansible/secrets/staging.enc --vault-password-file ansible/secrets/staging.vault-password -t deploy ansible/playbook.yml -e env_branch=fm/client-ops

PLAY [web] *********************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************
ok: [web1-staging]

TASK [common : Ensure nginx is running] ****************************************************************************************************************************************************************
ok: [web1-staging]

TASK [common : Ensure supervisor is running] ***********************************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Ensure Nginx site configuration is up to date] *********************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure Nginx site is default] **************************************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Git checkout fm/client-ops] ****************************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Check if .env exists] **********************************************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Ensure .env exists] ************************************************************************************************************************************************************************
skipping: [web1-staging]

TASK [web : Ensure database URL is present in .env file] ***********************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Get Python binary path] ********************************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Set python_bin_path fact] ******************************************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Ensure Python dependencies are installed] **************************************************************************************************************************************************
ok: [web1-staging]

TASK [web : Ensure Node dependencies are installed] ****************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure build is up to date] ****************************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure migrations are up to date] **********************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure supervisor configuration is up to date] *********************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure client is running] ******************************************************************************************************************************************************************
changed: [web1-staging]

TASK [web : Ensure server is running] ******************************************************************************************************************************************************************
ok: [web1-staging]

RUNNING HANDLER [common : reload nginx] ****************************************************************************************************************************************************************
changed: [web1-staging]

RUNNING HANDLER [common : reload supervisor] ***********************************************************************************************************************************************************
changed: [web1-staging]

PLAY RECAP *********************************************************************************************************************************************************************************************
web1-staging               : ok=19   changed=10   unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   

make[1]: Leaving directory '/home/florimond/dev/prj-catalogue/catalogage-donnees/ops'
```

Résultat : http://51.159.187.137/

* :x: Accéder à `/api` renvoie vers `/docs` (manque le préfixe) -> préfixe pas pris en compte par l'API (pb de configuration)
  * On le voie aussi en accédant à `/api/docs` directement : on obtient une erreur disant que `/openapi.json` n'a pas été trouvé (logique : il se trouve désormais sur `/api/openapi.json`)

=> Je viens de push un commit, mais je crois qu'il faudrait que je gère le "déplacement de l'API sur `/api/` en prod" dans une PR bien à part.

florimondmanca commented on client/package.json:
```diff
@@ -26,7 +27,7 @@
     "@typescript-eslint/parser": "^4.31.1",
     "cypress": "^9.4.1",
-    "c8": "^7.11.0",
+    "dotenv": "^16.0.0",
```
@magopian Je sais pas trop pourquoi, mais le `npm install --save-dev dotenv` a retiré `c8`. Ça n'a pas l'air de poser de pb.

magopian replied:
euh ... c8 est utilisé pour le coverage par vitest... tu peux quand même valider au cas où que le coverage est toujours opérationnel ?.

florimondmanca replied:
D'après mon run en local et les logs GitHub Actions, oui https://github.com/etalab/catalogage-donnees/runs/5055191759?check_suite_focus=true

florimondmanca commented:
@magopian J'ai testé de déployer sur staging. Il y a un problème avec le `node ./build`, ça réagit pas bien à un reload de Supervisor.

Sur l'instance, lors du deploy, j'obtenais une erreur bizarre du type `/_app/blabla.js not found` alors que le fichier était bien présent sur le serveur. Le `supervisorctl status client` me montrait que l'app avait planté. Les logs du client disaient "Address already in use", et pour cause : il y avait toujours bien l'ancien serveur sur le port 3000 (un petit `netstat -lp | grep 3000`). Après avoir kill cet ancien process, le client a redémarré et plus d'erreur.

En l'absence de doc sur le déploiement par SvelteKit j'ai utilisé ce fichier `build/index.js`, mais le SIGKILL envoyé par Supervisor au reload n'a pas l'air de lui faire grand effet. Je ne sais pas s'il faut ajouter quelque chose, ou si c'est "juste" un bug. J'en parlerai à M. SvelteKit demain...

magopian commented:
Eh ben... belle investigation !
Après une recherche google je tombe sur des infos de ce genre : 
- https://lisk.com/blog/development/why-we-stopped-using-npm-start-child-processes
- https://github.com/npm/npm/issues/4603

La solution semble d'être d'utiliser la commande node directement sans passer par `npm start`, ou d'avoir une magouille du genre https://github.com/lerna/lerna/issues/2284#issuecomment-895234817 (un script shell qui s'occupe de kill les process).

florimondmanca commented:
La solution semble converger, je lis ça ici aussi : https://stackoverflow.com/questions/46800198/how-can-supervisord-restart-the-npm-start-command-successfully

florimondmanca commented:
OK, testé sur la VM puis sur staging : `command=node ./build` lance bien 1 seul process que Supervisor gère correctement lors de l'update/restart. :+1:

florimondmanca commented:
Allez hop, voyons voir ce que ça donne à l'usage !
